neuralop/models/fino.py

```diff
@@ -0,0 +1,112 @@
+    '''
+    fft_type = {'sph', 'norm'}, if 'sph' it uses the speherical Fourier Transform 
+    '''
+    def __init__(self, in_channels, out_channels, n_modes, incremental_n_modes=None, bias=True,
+                 n_layers=1, separable=False, output_scaling_factor=None,
+                 rank=0.5, factorization='cp', implementation='reconstructed', 
+                 fixed_rank_modes=False, joint_factorization=False, decomposition_kwargs=dict(),
+                 init_std='auto', fft_norm='forward', fft_type = 'sht',  sht_nlat = 180, sht_nlon = 360,\
+                 sht_grid="legendre-gauss", isht_grid="legendre-gauss", sht_norm="backward", frequency_mixer = False):
+        super().__init__(in_channels, out_channels, n_modes, incremental_n_modes, bias,
+                 n_layers, separable, output_scaling_factor,
+                 rank, factorization, implementation, 
+                 fixed_rank_modes, joint_factorization, decomposition_kwargs,
+                 init_std, fft_norm)
+
+        #self.shared = shared
+        
+        hm1, hm2 = self.half_n_modes[0], self.half_n_modes[1]
+        if self.factorization is None or not frequency_mixer:
+            self.W1 = nn.Parameter(torch.empty(hm1, hm2, hm1, hm2, dtype = torch.cfloat))
+            self.W2 = nn.Parameter(torch.empty(hm1, hm2, hm1, hm2, dtype = torch.cfloat))
+            self.reset_parameter()
+        self.sht_grid = sht_grid
+        self.isht_grid = isht_grid
+        self.sht_norm = sht_norm
+        self.fft_type = fft_type
+        self.frequency_mixer = frequency_mixer
+        
+        ####
+        # This following line might have a version dependent nature
+        ####
+        if self.output_scaling_factor is not None:
+            out_nlat = int(round(sht_nlat * self.output_scaling_factor[0][0]))
+            out_nlon = int(round(sht_nlon * self.output_scaling_factor[0][1]))
+        
+
+        if fft_type == 'sht':
+            self.forward_fft = th.RealSHT(sht_nlat, sht_nlon, grid = self.sht_grid, norm= self.sht_norm)
+            self.inverse_fft = th.InverseRealSHT(out_nlat, out_nlon, grid = self.isht_grid, norm= self.sht_norm)
+            
+
+    def reset_parameter(self):
+        # Initial model parameters.
+        scaling_factor = 1/(self.in_channels * self.half_n_modes[0]*self.half_n_modes[1])**0.5
+        torch.nn.init.normal_(self.W1, mean=0.0, std=scaling_factor)
+        torch.nn.init.normal_(self.W2, mean=0.0, std=scaling_factor)
+    
+    def mode_mixer(self, input, weights):
+        return torch.einsum("bimn,mnop->biop", input, weights)
+    
+    def forward(self, x, indices=0, output_shape = None):
+        batchsize, channels, height, width = x.shape
+
+        if self.fft_type == 'sht':
+            if self.forward_fft.nlat != x.shape[-2] or self.forward_fft.nlon != x.shape[-1]:
+                self.forward_fft = th.RealSHT(x.shape[-2], x.shape[-1], grid = self.sht_grid, norm= self.sht_norm).to(x.device)
```

Comment:
Let's reuse a similar caching as in the Spherical Conv -- we could also write a small wrapper on SHT to do the caching. Let's debate with @bonevbs